modules/tournament/src/main/Tournament.scala

```diff
@@ -102,27 +111,135 @@ case class Tournament(
 
   def pairingsClosed = secondsToFinish < pairingsClosedSeconds
 
+  //start at 0 as it's actaully an index for medley variants in front end
   def medleyRound: Option[Int] =
-    if (isStarted) medleyMinutes.map(_ * 60).map((nowSeconds - startsAt.getSeconds).toInt / _)
-    else None
+    if (isStarted) {
+      medleyIntervalSeconds
+        .map(
+          _.map { var s = 0; t => { s += t; s } }
+            .filter(_ <= ((nowSeconds - startsAt.getSeconds).toInt))
+            .size
+        )
+    } else None
+
+  def finalMedleyVariant: Boolean = medleyRound.getOrElse(-99) == medleyNumIntervals.getOrElse(0) - 1
+
+  def medleyIntervalSeconds: Option[List[Int]] =
+    if (medleyBalanceIntervals) {
+      val times: Option[List[Int]] =
+        medleySpeedFactors.map(f => f.map(s => (s * (minutes / f.sum) * 60).toInt))
+      val extra               = times.map(minutes * 60 - _.sum).getOrElse(0)
+      val firstLastBonus: Int = math.min(clock.limitSeconds / 3, 120)
+      // take time from first variant and give to last
+
+      times.map(t =>
+        t.take(1).map(v => v - firstLastBonus) ::: t.drop(1).take(t.size - 2) :::
+          t.drop(t.size - 1).map(v => v + extra + firstLastBonus)
+      )
+    } else {
+      medleyNumIntervals.map(List.fill(_)(medleyMinutes.getOrElse(0) * 60))
+    }
+
+  def medleySpeedFactors: Option[List[Double]] =
+    medleyVariantsInTournament.map(_.map(v => medleySpeedChoice.get(v.key).getOrElse(1.0)))
+
+  def medleySpeedChoice: Map[String, Double] = medleyVariants.fold(medleyVariantSpeeds)(v =>
```

Comment:
this surely wants to be a lazy val.
combined with the function above for every variant in the tournament we're re-working out the style of medley and therefore which set of speeds to use, when we only need to do it once

Reply:
okay